### PR TITLE
experiment(ir): remove intermediate expressions

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
can we keep the original values around for a while - this is a major breaking change

**Comment by kszucs** [comment]
@jreback this is just a prototype where we can experiment with pattern matching on operation trees using matchpy

File: ibis/expr/operations/generic.py [hunk 1]
```diff
@@ -273,7 +274,7 @@ def root_tables(self):
 
 
 @public
-class NullLiteral(Literal):
+class NullLiteral(Literal, Singleton):
```
Comment (kszucs) [inline ibis/expr/operations/generic.py:277]: Ohh how handy it is!

**Comment by kszucs** [comment]
In order to allow proper pattern matching on ibis operations we would need to accept three kinds of arguments:
1. Expressions arriving from the standard usage of the API
2. `matchpy.Wildcard` and perhaps other `matchpy.Symbol` objects
3. ibis operations so we can substitute operations without converting to expression before construction 

These three ideally should be achieved by rewriting the current rules to return with ibis operations instead of expressions.

**Comment by kszucs** [comment]
Closing in favor of https://github.com/ibis-project/ibis/pull/4135

**Pull request closed without merging**